File: java/src/main/java/io/github/hc1839/xbelmark/xslt/ext/DateTime.java
package io.github.hc1839.xbelmark.xslt.ext;

import java.time.DateTimeException;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoField;
import java.time.temporal.TemporalAccessor;

/**
 *  Conversions of date/time formats.
 */
public class DateTime {
  /**
   *  Converts `xs:dateTime` or `xs:date` to seconds since epoch.
   *
   *  @param input
   *    Date/time string in `xs:dateTime` or `xs:date` format.
   *
   *  @return
   *    Seconds since epoch.
   */
  public static long dateTimeToUnix(String input) throws DateTimeException {
    TemporalAccessor temporal;
    LocalTime time;
    try {
      temporal = DateTimeFormatter.ISO_DATE_TIME.parse(input);
      time = LocalTime.from(temporal);
    } catch (Exception e) {
      temporal = DateTimeFormatter.ISO_DATE.parse(input);
      time = LocalTime.MIDNIGHT;
    }
    LocalDate date = LocalDate.from(temporal);
    ZoneId zone = temporal.isSupported(ChronoField.OFFSET_SECONDS)
                  ? ZoneId.from(temporal)
                  : ZoneId.systemDefault();
    return ZonedDateTime.of(date, time, zone).toEpochSecond();
  }
}

File: dotnet/src/main/csharp/Xbelmark/Xslt/Ext/DateTime.cs
using System.Globalization;
using System;

namespace Xbelmark.Xslt.Ext;

/// <summary>
/// Conversions of date/time formats.
/// </summary>
public class DateTime
{
    /// <summary>
    /// Converts `xs:dateTime` or `xs:date` to seconds since epoch.
    /// </summary>
    /// <param name="input">
    /// Date/time string in `xs:dateTime` or `xs:date` format.
    /// </param>
    /// <returns>
    /// Seconds since epoch.
    /// </returns>
    public long dateTimeToUnix(string input)
    {
        DateTimeOffset dto;
        try
        {
            dto = DateTimeOffset.ParseExact(
                input, "O", null, DateTimeStyles.RoundtripKind);
        }
        catch (Exception)
        {
            dto = DateTimeOffset.ParseExact(
                input, "yyyy'-'MM'-'dd'T'HH':'mm':'sszzz", null);
        }
        return dto.ToUnixTimeSeconds();
    }
}
